rfcs/20210219-cache-compiled-gpu-kernels/README.md

```diff
@@ -0,0 +1,101 @@
+which will provide required API
+
+For the models that have small execution time, for example DLRM or BERT the
+overhead becomes critical. All optimizations that can be done at oneDNN side is
+completed. They allowed to get rid of overhead for models like ResNet-50 (MB 1024),
+but they didn't solve the problem.
+
+The only viable option is to get rid of the overhead completely. The proposal is
+to switch to caching compiled kernel rather than their binary representations.
+
+## How to deal with the unloading libraries issue?
+
+In general, it's seem unlikely to run into the issue because:
+1. According to the SYCL integration design, frameworks store GPU device and context
+abstractions globally. If that was an issue they wouldn't have done that
+2. Benchdnn and Gtests have global engine which contains GPU device and context
+abstractions and no issues so far
+
+If it becomes an issue, the workaround will be clearing the primitive cache by
+calling  `dnnl::set_primitive_cache_capacity(0);` before exit on the frameworks' side.
+
+Also, SYCL is trying to handle such situations properly. For example on Linux,
+they use `__attribute__((destructor))` with low priority to make sure that the libraries
+that do not use the constructor with lower priority will be unloaded before the SYCL
+Runtime. [Here](https://github.com/intel/llvm/blob/sycl/sycl/doc/GlobalObjectsInRuntime.md)
+is the document that describes what they do to prevent such issues.
+
+## Benefits from caching compiled kernels
+
+There is a number of benefits:
+1. Eliminate overhead on creating kernels from binary and copying scales
+2. Significantly reduce code complexity
+    * No need in having resource for GPU
+    * No need in having complex abstractions that can contain either binary or kernel
+3. Everything can be stored in the primitive
+
+## DPC++ cache
+
+Not directly related to the proposal but DPC++ is going to implement so-called
+["persistent cache"](https://github.com/intel/llvm/blob/sycl/sycl/doc/KernelProgramCache.md)
+which is storing some internal resources in a data storage. It means that time
+on first primitive creation for DPC++ kernels can be reduced significantly if
+there is a corresponding resource in the cache in the data storage.
+If oneDNN caches compiled kernels and the first iterations benefits from the
+"persistent cache" then time on primitive creation will be reduced as much as it's
+technically possible.
+
+
+## Primitive cache key
+
+The key should be extended to take into account GPU context. Currently, there is
+`device_id_t` which is basically an alias for `std::tuple`. The key doesn't store
+GPU devices themselves but a few numbers that can be used to distinguish one
+GPU device from another.
+
+In order to switch to caching compiled kernels there will be introduced `engine_id_t`
+which will be responsible for holding common information e.g. engine kind,
+runtime kind and index. And CPU or GPU specific information too, e.g. GPU device
```

Comment:
how would that work with interop APIs (in particular for the index part)?

Reply:
The index should be default initialized (e.g. -1) when an interop API is used to created an engine. But in general it looks like we don't need index to compare engines.

The idea is to have `engine_id_t` which is in other words a cachable engine with defined comparison operators.
For example:
```cpp
// User side
engine eng1(gpu, dev, ctx)
engine eng2(gpu, 0)

// Inside the library
engine_id_t eng_id1(eng1);
engine_id_t eng_id2(eng2);

// eq is false
bool eq = eng_id1 == eng_id2;
```